### PR TITLE
DriveUntilHitHub

### PR DESCRIPTION
**Pull request by yutytuty**

### CODE REVIEW

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -49,6 +46,11 @@ public void robotInit() {
 
         rootNamespace = new RootNamespace("robot namespace");
         rootNamespace.putData("intake cargo", new IntakeCargo());
+        rootNamespace.putData("drive forward", new DriveArcade(drivetrain, 0.5, 0));
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:49]: remove

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -49,6 +46,11 @@ public void robotInit() {
...
+        rootNamespace.putData("drive backward", new DriveArcade(drivetrain, -0.5, 0));
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:50]: ditto

File: src/main/java/frc/robot/Robot.java [hunk 3]
```diff
@@ -49,6 +46,11 @@ public void robotInit() {
...
+        rootNamespace.putData("drive until hit hub", new DriveUntilHitHub(drivetrain));
+        rootNamespace.putNumber("left talon current", drivetrain.getLeftTalon()::getStatorCurrent);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:52]: remove

File: src/main/java/frc/robot/Robot.java [hunk 4]
```diff
@@ -49,6 +46,11 @@ public void robotInit() {
...
+        rootNamespace.putNumber("right talon current", drivetrain.getRightTalon()::getStatorCurrent);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:53]: remove

File: src/main/java/frc/robot/Robot.java [hunk 5]
```diff
@@ -65,13 +67,6 @@ public boolean isFinished() {
                 return false;
             }
         });
-
-        Supplier<Double> turn = rootNamespace.addConstantDouble("turn value", 0.3);
-        Supplier<Double> move = rootNamespace.addConstantDouble("move value", 0.3);
-        rootNamespace.putData("turn robot", new DriveArcade(drivetrain, () -> 0.0, turn));
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:71]: add all of those

File: src/main/java/frc/robot/commands/DriveUntilHitHub.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    public boolean isFinished() {
+        return Math.abs(drivetrain.getRightTalon().getStatorCurrent()) > STALL_CURRENT.get() &&
+                Math.abs(drivetrain.getLeftTalon().getStatorCurrent()) > STALL_CURRENT.get();
+    }
+
+    public static void periodic() {
+        rootNamespace.update();
```
Comment (tal-sitton) [inline src/main/java/frc/robot/commands/DriveUntilHitHub.java:36]: what are you updating?

File: src/main/java/frc/robot/commands/DriveUntilHitHub.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ * Drives the robot forward until it hits the hub. Robot needs to be pre aligned to the hub.
+ */
+public class DriveUntilHitHub extends DriveArcade {
+
+    private static final RootNamespace rootNamespace = new RootNamespace("drive until hit hub");
+
+    /**
+     * A value between the stall current and running current of the motors
+     */
+    private static final Supplier<Double> STALL_CURRENT = rootNamespace.addConstantDouble("stall current", 0);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/commands/DriveUntilHitHub.java:19]: add the real value to it

File: src/main/java/frc/robot/commands/DriveUntilHitHub.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    private static final Supplier<Double> MOVEMENT_SPEED = rootNamespace.addConstantDouble("movement speed", 0);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/commands/DriveUntilHitHub.java:20]: ditto

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 9]
```diff
@@ -39,56 +39,64 @@
...
     private final PigeonWrapper pigeon;
+    private final WPI_TalonSRX leftTalon, rightTalon;
     private final Encoder leftEncoder, rightEncoder;
 
-    private final Supplier<Double> kPGyro = gyroPIDNamespace.addConstantDouble("kP", 0.017);
-    private final Supplier<Double> kIGyro = gyroPIDNamespace.addConstantDouble("kI", 0.0028);
-    private final Supplier<Double> kDGyro = gyroPIDNamespace.addConstantDouble("kD", 0);
-    private final Supplier<Double> toleranceGyro = gyroPIDNamespace.addConstantDouble("tolerance", 5);
-    private final Supplier<Double> waitTimeGyro = gyroPIDNamespace.addConstantDouble("wait time", 0.5);
+    private final Supplier<Double> kPGyro = gyroPIDNamespace.addConstantDouble("gyro kP", 0);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:45]: why 0?

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 10]
```diff
@@ -39,56 +39,64 @@
...
+    private final Supplier<Double> kIGyro = gyroPIDNamespace.addConstantDouble("gyro kI", 0);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:46]: ditto

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 11]
```diff
@@ -39,56 +39,64 @@
...
+    private final Supplier<Double> kDGyro = gyroPIDNamespace.addConstantDouble("gyro kD", 0);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:47]: ditto

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 12]
```diff
@@ -39,56 +39,64 @@
     private static final Supplier<Double> leftCorrection = corrections.addConstantDouble("left correction", 1);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:39]: ok you changed so much in this class I'm not going to even comment on everything,
just change all of these to the right things in `autonomous-testing`

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/DriveUntilHitHub.java (hunk 6 above)
Comment (yutytuty) [reply src/main/java/frc/robot/commands/DriveUntilHitHub.java:36]: The values of the namespace

File: src/main/java/frc/robot/Robot.java [hunk 13]
```diff
@@ -8,10 +8,10 @@
 import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
 import com.spikes2212.dashboard.RootNamespace;
 import edu.wpi.first.wpilibj.TimedRobot;
+import edu.wpi.first.wpilibj.Timer;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Robot.java:11]: ?

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 14]
```diff
@@ -70,6 +71,8 @@
     public static Drivetrain getInstance() {
         if (instance == null) {
             WPI_TalonSRX pigeonTalon = new WPI_TalonSRX(RobotMap.CAN.PIGEON_TALON);
+            WPI_TalonSRX leftTalon = new WPI_TalonSRX(RobotMap.CAN.DRIVETRAIN_LEFT_TALON_1);
+            WPI_TalonSRX rightTalon = new WPI_TalonSRX(RobotMap.CAN.DRIVETRAIN_RIGHT_TALON_1);
             instance = new Drivetrain(new BustedMotorControllerGroup(
                     leftCorrection,
                     new WPI_TalonSRX(RobotMap.CAN.DRIVETRAIN_LEFT_TALON_1),
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:78]: you have already created this talon

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 15]
```diff
@@ -80,15 +83,20 @@ public static Drivetrain getInstance() {
                             new WPI_TalonSRX(RobotMap.CAN.DRIVETRAIN_RIGHT_TALON_1),
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:83]: ditto

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/DriveUntilHitHub.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+    private static final RootNamespace rootNamespace = new RootNamespace("drive until hit hub");
+
+    /**
+     * A value between the stall current and running current of the motors
+     */
+    private static final Supplier<Double> STALL_CURRENT = rootNamespace.addConstantDouble("stall current", 15);
+    private static final Supplier<Double> MOVEMENT_SPEED = rootNamespace.addConstantDouble("movement speed", 0.5);
+
+    private final Drivetrain drivetrain;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/DriveUntilHitHub.java:22]: you already have a protected `tankDrivetrain` in the superclass, why create another one here?

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java [hunk 17]
```diff
@@ -50,6 +51,13 @@ public void robotInit() {
         rootNamespace = new RootNamespace("robot namespace");
         rootNamespace.putData("intake cargo", new IntakeCargo());
 
+        Supplier<Double> turn = rootNamespace.addConstantDouble("turn value", 0.3);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:54]: put all the namespace things **after** the `setDefaultCommand`

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 18]
```diff
@@ -70,25 +71,32 @@
...
         }
         return instance;
     }
 
-    private Drivetrain(MotorControllerGroup leftMotors, BustedMotorControllerGroup rightMotors, WPI_TalonSRX pigeonTalon) {
+    private Drivetrain(MotorControllerGroup leftMotors, BustedMotorControllerGroup rightMotors, WPI_TalonSRX pigeonTalon,
+                        WPI_TalonSRX leftTalon, WPI_TalonSRX rightTalon) {
         super("drivetrain", leftMotors, rightMotors);
         this.pigeon = new PigeonWrapper(pigeonTalon);
+        this.leftTalon = leftTalon;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:97]: leftTalon = pigeonTalon

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Review by tal-sitton — APPROVED** [review]

**Pull request merged**